File: src/factories/internal/metadata/iterate_metadata_tuple.ts
import ts from "typescript";

import { Metadata } from "../../../schemas/metadata/Metadata";
import { MetadataTuple } from "../../../schemas/metadata/MetadataTuple";
import { MetadataTupleType } from "../../../schemas/metadata/MetadataTupleType";

import { ArrayUtil } from "../../../utils/ArrayUtil";

import { MetadataCollection } from "../../MetadataCollection";
import { MetadataFactory } from "../../MetadataFactory";
import { emplace_metadata_tuple } from "./emplace_metadata_tuple";

export const iterate_metadata_tuple =
    (checker: ts.TypeChecker) =>
    (options: MetadataFactory.IOptions) =>
    (collection: MetadataCollection) =>
    (errors: MetadataFactory.IError[]) =>
    (
        meta: Metadata,
        type: ts.TupleType,
        explore: MetadataFactory.IExplore,
    ): boolean => {
        if (!checker.isTupleType(type)) return false;

        const tupleType: MetadataTupleType = emplace_metadata_tuple(checker)(
            options,
        )(collection)(errors)(type, meta.nullable, explore);
        ArrayUtil.add(
            meta.tuples,
            MetadataTuple.create({
                type: tupleType,
                tags: [],
            }),
            (elem) => elem.type.name === tupleType.name,
        );
        return true;
    };

File: packages/typia/native/core/factories/internal/metadata/iterate_metadata_tuple.go
package metadata

import (
  nativechecker "github.com/microsoft/typescript-go/shim/checker"
  schemametadata "github.com/samchon/typia/packages/typia/native/core/schemas/metadata"
)

func Iterate_metadata_tuple(props IMetadataIteratorProps) bool {
  if props.Checker == nil || nativechecker.IsTupleType(props.Type) == false {
    return false
  }

  tupleType := Emplace_metadata_tuple(props)
  for _, elem := range props.Metadata.Tuples {
    if elem.Type.Name == tupleType.Name {
      return true
    }
  }
  props.Metadata.Tuples = append(props.Metadata.Tuples, schemametadata.MetadataTuple_create(schemametadata.MetadataTuple{
    Type: tupleType,
    Tags: [][]schemametadata.IMetadataTypeTag{},
  }))
  return true
}
